compilation error for displaying map of references of generic types

### Describe the bug

The following code is getting a compilation error:

```
import datatypes
fn main() {
        mut m := map[string]&datatypes.Queue[i64]{}
        println("${m}")
}
```

### Expected Behavior

The code should compile cleanly.

### Current Behavior

v fails with the following errors:
```
/tmp/v_5014/map.647552516261589458.tmp.c:2214: error: cannot convert 'struct datatypes__Queue_T_i64 *' to 'struct datatypes__Queue_T_i64'
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang


```

### Reproduction Steps

v run bug.v

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 3c23a7e

### Environment details (OS name and version, etc.)

OS: linux, "CentOS Linux release 7.9.2009 (Core)"
Processor: 8 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E3-1231 v3 @ 3.40GHz
CC version: cc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)

V full version: V 0.3.3 5454562.3c23a7e

Git vroot status: weekly.2023.09-25-g3c23a7eb-dirty
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3-dirty

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -759,7 +759,8 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 		tmp_str := str_intp_rune('${elem_str_fn_name}(*(${val_styp}*)DenseArray_value(&m.key_values, i))')
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${tmp_str});')
 	} else {
-		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${elem_str_fn_name}(*(${val_styp}*)DenseArray_value(&m.key_values, i)));')
+		ptr_str := '*'.repeat(val_typ.nr_muls())
+		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${elem_str_fn_name}(*${ptr_str}(${val_styp}*)DenseArray_value(&m.key_values, i)));')
 	}
 	g.auto_str_funcs.writeln('\t\tif (i != m.key_values.len-1) {')
 	g.auto_str_funcs.writeln('\t\t\tstrings__Builder_write_string(&sb, _SLIT(", "));')

--- a/vlib/v/tests/map_reference_value_test.v
+++ b/vlib/v/tests/map_reference_value_test.v
@@ -1,3 +1,5 @@
+import datatypes
+
 struct Foo {
 	bar string
 }
@@ -14,4 +16,10 @@ fn test_map_reference_value() {
 	println(m2)
 
 	assert true
+}
+
+fn test_map_reference_value2() {
+	mut m := map[string]&datatypes.Queue[i64]{}
+	println('${m}')
+	assert '${m}' == '{}'
 }

